File: 0334_increasingTriplet.py
class Solution:
    def increasingTriplet(self, nums: List[int]) -> bool:
        dp = []
        for num in nums:
            index = bisect.bisect_left(dp, num)
            if index == len(dp):
                dp.append(num)
            else:
                dp[index] = num
            if len(dp) == 3:
                return True
        return False
        
# Alternative solution

class Solution1:
    def increasingTriplet(self, nums: List[int]) -> bool:
        a, b, c = None, None, None
        for num in nums:
            if a == None:
                a = num
            elif b == None:
                if num <= a:
                    a = num
                else:
                    b = num
            elif c == None:
                if num <= a:
                    a = num
                elif num <= b:
                    b = num
                else:
                    return True
        return False

File: 0334_increasingTriplet.java
class Solution {
    public boolean increasingTriplet(int[] nums) {
        int a = Integer.MIN_VALUE;
        int b = Integer.MIN_VALUE;
        int c = Integer.MIN_VALUE;
        for (int i = 0; i < nums.length; i++) {
            if (a == Integer.MIN_VALUE) {
                a = nums[i];
            } else if (b == Integer.MIN_VALUE) {
                if (nums[i] <= a) {
                    a = nums[i];
                } else {
                    b = nums[i];
                }
            } else if (c  == Integer.MIN_VALUE) {
                if (nums[i] <= a) {
                    a = nums[i];
                } else if (nums[i] <= b) {
                    b = nums[i];
                } else {
                    return true;
                }
            }
        }
        return false;
    }
}
